starboard/common/socket.cc

```diff
@@ -51,121 +57,473 @@ bool GetLocalhostAddress(SbSocketAddressType address_type,
+      struct sockaddr_in* addr_in = (struct sockaddr_in*)&saddr;
+      out_address->type = kSbSocketAddressTypeIpv4;
+      out_address->port = addr_in->sin_port;
+      memcpy(&out_address->address, &addr_in->sin_addr, sizeof(struct in_addr));
+    }
+#if SB_HAS(IPV6)
+    if (saddr.sa_family == AF_INET6) {
+      struct sockaddr_in6* addr_in6 = (struct sockaddr_in6*)&saddr;
+      out_address->type = kSbSocketAddressTypeIpv6;
+      out_address->port = addr_in6->sin6_port;
+      memcpy(&out_address->address, &addr_in6->sin6_addr,
+             sizeof(struct in6_addr));
+    }
+#endif
+  }
+#else
   return SbSocketGetLocalAddress(socket_, out_address);
+#endif  // SB_API_VERSION >= 16
 }
 
 bool Socket::SetBroadcast(bool value) {
+#if SB_API_VERSION >= 16
+  int true_val = value ? 1 : 0;
+  if (::setsockopt(socket_, SOL_SOCKET, SO_BROADCAST, &true_val, value) == 0) {
+    return true;
+  }
+  return false;
+#else
   return SbSocketSetBroadcast(socket_, value);
+#endif  // SB_API_VERSION >= 16
 }
 
 bool Socket::SetReuseAddress(bool value) {
+#if SB_API_VERSION >= 16
+  return true;
+  const int on = value ? 1 : 0;
+  if (::setsockopt(socket_, SOL_SOCKET, SO_REUSEADDR, &on, sizeof(on)) == 0) {
+    return true;
+  }
+  return false;
+#else
   return SbSocketSetReuseAddress(socket_, value);
+#endif  // SB_API_VERSION >= 16
 }
 
 bool Socket::SetReceiveBufferSize(int32_t size) {
+#if SB_API_VERSION >= 16
+  if (::setsockopt(socket_, SOL_SOCKET, SO_RCVBUF, &size, sizeof(size)) == 0) {
+    return true;
+  }
+  return false;
+#else
   return SbSocketSetReceiveBufferSize(socket_, size);
+#endif  // SB_API_VERSION >= 16
 }
 
 bool Socket::SetSendBufferSize(int32_t size) {
+#if SB_API_VERSION >= 16
+  return true;
+  // skip setsockopt for POSIX socket, otherwise send() will hang
+  if (::setsockopt(socket_, SOL_SOCKET, SO_SNDBUF, &size, sizeof(size)) != 0) {
+    return false;
+  }
+#else
   return SbSocketSetSendBufferSize(socket_, size);
+#endif  // SB_API_VERSION >= 16
 }
 
 bool Socket::SetTcpKeepAlive(bool value, int64_t period) {
+#if SB_API_VERSION >= 16
+  const int on = value ? 1 : 0;
+  if (::setsockopt(socket_, IPPROTO_TCP, TCP_NODELAY, &on, sizeof(on)) != 0) {
+    return false;
+  }
+  if (value) {
+    int result = 0;
+    result |= ::setsockopt(socket_, SOL_SOCKET, SO_KEEPALIVE, &on, sizeof(on));
+
+#if defined(__APPLE__)
+    // In tvOS, TCP_KEEPIDLE and SOL_TCP are not available.
+    // For reference:
+    // https://stackoverflow.com/questions/15860127/how-to-configure-tcp-keepalive-under-mac-os-x
+    result |= ::setsockopt(socket_, IPPROTO_TCP, TCP_KEEPALIVE, &period,
+                           sizeof(period));
+#elif !defined(_WIN32)
+    // In Windows, the SOL_TCP and TCP_KEEPIDLE options are not available.
+    // For reference:
+    // https://stackoverflow.com/questions/8176821/how-to-set-the-keep-alive-interval-for-winsock
+    result |=
+        ::setsockopt(socket_, SOL_TCP, TCP_KEEPIDLE, &period, sizeof(period));
```

Comment:
Could you please try replacing all the SOL_TCP with IPPROTO_TCP to see if it works?

Reply:
I think Socket wrapper is not needed in SB16. Therefore we don't need to implement these POSIX calls for the wrapper here.